src/caffe/test/test_gradient_based_solver.cpp

```diff
@@ -483,6 +495,29 @@ TYPED_TEST(AdaGradSolverTest, TestLeastSquaresUpdateWithEverythingAccum) {
 }
 
 template <typename TypeParam>
+class AdamSolverTest : public GradientBasedSolverTest<TypeParam> {
+  typedef typename TypeParam::Dtype Dtype;
+
+ protected:
+  virtual void InitSolver(const SolverParameter& param) {
+    this->solver_.reset(new AdamSolver<Dtype>(param));
+  }
+  virtual SolverParameter_SolverType solver_type() {
+    return SolverParameter_SolverType_ADAM;
+  }
+};
+
+TYPED_TEST_CASE(AdamSolverTest, TestDtypesAndDevices);
+TYPED_TEST(AdamSolverTest, TestAdamLeastSquares) {
+  typedef typename TypeParam::Dtype Dtype;
+  const Dtype kLearningRate = 1.0;
+  const Dtype kMomentum = 0.9;
+  // momentum2 is already set to 0.999
+  const Dtype kWeightDecay = 0.0;
+  this->TestLeastSquaresUpdate(kLearningRate, kWeightDecay, kMomentum);
+}
+
```

Comment:
Please add test cases to 4 more test cases: 

* `TestAdamLeastSquaresUpdateLROneTenth` (where you set `kLearningRate = 0.1`).
* `TestAdamLeastSquaresUpdateWithWeightDecay` (where you may set `kWeightDecay = 0.5`).
* `TestAdamLeastSquaresUpdateWithEverything` (where you may set `kWeightDecay = 0.1` and `kNumIters = 4` and loop for `kNumIters` times)
* `TestLeastSquaresUpdateWithEverythingAccum` (where you may set `kNumIters = 4` and `kIterSize = 2` and use `CheckAccumulation`)

You may take a look at those AdaGradSolverTest for details.